### DISCUSSION TITLE
Adding before\prepend\skip_before_render controller hooks

### DISCUSSION THREAD

**Discussion by aarona** [rails/rails#53358] [Ideas]
## Background:
I've been using a gem for a long time where I work that we use for running methods after a controller action completes and before our views are rendered. The main use case we have is for doing some authentication checks. I'm sure there are other use cases for these hooks as well.

This gem's origin is from [16 years ago with Rails 2 ](https://github.com/xtoddx/before_render) but has been forked many times over by other people specifically to support newer versions of Rails until you have [what I use today](https://github.com/aarona/before_render) which currently supports Rails 7. You can see that the [combined downloads of this gem in all its generational iterations](https://rubygems.org/search?query=before_render) is over 220k which I believe to be pretty significant.

[Here's our current version of this gem](https://github.com/aarona/before_render).

## Idea:
After having to modify our personal version to upgrade its Rails dependencies (yet again) I discussed the idea of creating a PR for Rails with everyone at work. Someone mentioned that there must be a good reason why no one has already done this (citing possible 'conceptual disagreements'). However, after doing a little research under Issues, Pull requests and Discussions I didn't find anything related to adding or not adding controller hooks like this so I asking for thoughts about this here.

## Why I think this is worth adding:
1. **The code has already been written:** No need to write code for this feature. Simply add the existing code into Rails.
2. **The code is stable:** This code hasn't changed much in a long time with exceptions to explicitly tell it to support newer versions of Rails. There's nothing really complicated -- It's clean, simple and feature complete.
3. **It's lightweight**: The meat and potatoes of this gem is less than 60 lines of code. Everything else is gem boiler plate stuff.
4. **It already has spec tests:** It already has a bunch of tests that handle any cases you could throw at it. However feel free to [look at the tests specifically](https://github.com/aarona/before_render/tree/578164095b0d599a0cf9cc3466b306360c4f2a0d/lib/before_render) and see if something needs to be added.
5. **I'd be willing to create the PR myself:** I've contributed to open source already but I haven't contributed to Rails yet that I can remember. It would be a good experience for me and I'm essentially already bringing the solution to my own problem (and the many others who have also needed something like this feature in the past).

**Comment by jeromedalbert** [comment]
What is the benefit of doing authentication checks before_render instead of before_action?

**Reply by aarona** [reply]
> What is the benefit of doing authentication checks before_render instead of before_action?

Great question. It seems that our use case for this is so that we can force Pundit authorization to be called in the action so it does not render the reponse if we haven't checked that the user has access to that particular resource.

Here's an example of how we do this:

```ruby
module PunditAuth
  extend ActiveSupport::Concern
  include Pundit::Authorization

  included do
    # This is where the before_render hook would come into play.
    # Right now we need to include a gem to do this
    before_render :verify_authorized, except: :index
    before_render :verify_policy_scoped, only: :index
  end

  # more code...
end

class SomeController < ApplicationController
  include PunditAuth
  
  def show
    # Pundit authorization took place in the action so the verification test will pass
    authorize(...)
  end
  
  def index
    # not verifying policy scope here would raise an expected error
  end
end
```

Also, here's a [Stack Overflow thread](https://stackoverflow.com/questions/9281224/filter-to-execute-before-render-but-after-controller) where someone needs that same behavior where one of the solutions was the rails3_before_render gem. 

The next question might be _"Well why not put this check in an `after_action` hook?"_ The issue there is that redirection calls might have already occurred and it seems that the response is already rendered as well. I found [another StackOverflow example](https://stackoverflow.com/questions/68227095/how-to-use-after-action-callback-in-rails-controller) of someone needing to check if a user is "pro" (analogous to an authorization check) and it seemed that the solution for this person was to use a `before_action` hook. In our case however, we might need call `Pundit`'s `authorize` check in the middle of the action after a query has been made or something else has occurred.

**Comment by fillipeppalhares** [comment]
I've been searching for a hook like this for a while now. I understand it is particularly useful when you have a `concern` shared between many controllers, and you have another `concern` that only a few of them will use.

**Example use case**
---
```rb
# app/controllers/concerns/concern_utilized_by_many.rb

module ConcernUtilizedByMany
  def my_action
    @object = find_record.assign_attributes(set_params)

    # some code
  end
end
```

The `my_action` action is broadly used, but for some other entities, I would need to make a minor adjustment:

```rb
# app/controllers/concerns/concern_utilized_by_few.rb

module ConcernUtilizedByFew
  def my_action
    @object = find_record.assign_attributes(set_params)

    fill_object_address if params.find_key("address")
    
    # some code
  end
end
```

As you can see it, the way I've managed to achieve it was overriding `my_action`. Definitely not the best thing in the world. The way I see it, a `before_render` is useful when you are dealing with a lot of `concerns` and you need minor adjustments in each one of them, for the same action.

---

Yesterday, I found a way I could solve it more efficiently. There it is what I did:

```rb
# app/controllers/application_controller.rb

class ApplicationController < ActionController::Base
  define_callbacks :render

  def render(*args, &block)
    @__controller_instance_variables = instance_variables

    run_callbacks(:render) { super }
  end

  private

  def self.before_render(*args, only: nil, except: nil, &block)
    condition = lambda do |controller|
      action = controller.action_name.to_sym

      next false if only && !only.include?(action)
      next false if except && except.include?(action)

      true
    end

    set_callback(:render, :before, *args, if: condition, &block)
  end
end
```

I've created a custom hook called `before_render` and now it looks just like any other hook:

```rb
before_render :do_something, only: %i[new edit]
```

---

I think it is a useful tool, and it could already be packed in with Rails. Just like @aarona, I wonder why no one have done this before. That's my two cents on the matter. I look forward to discussing the idea. I'm also open to contribute to the code.

**Reply by anoiaque** [reply]
Thanks for that !
I just update for me !only.include?(action) by ![only].flatten.include?(action) if you declare for exemple without array, ie before_render, only: :create.
Idem for except.

**Reply by fillipeppalhares** [reply]
oh yes, great observation!
